<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.33">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>readme</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="README_files/libs/clipboard/clipboard.min.js"></script>
<script src="README_files/libs/quarto-html/quarto.js" type="module"></script>
<script src="README_files/libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="README_files/libs/quarto-html/popper.min.js"></script>
<script src="README_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="README_files/libs/quarto-html/anchor.min.js"></script>
<link href="README_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="README_files/libs/quarto-html/quarto-syntax-highlighting-ea385d0e468b0dd5ea5bf0780b1290d9.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="README_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="README_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="README_files/libs/bootstrap/bootstrap-81267100e462c21b3d6c0d5bf76a3417.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent quarto-light">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">




<section id="nanomb-oxford-nanopore-16s-pipeline-container-first" class="level1">
<h1><img src="graphics/logo.png" class="img-fluid" width="70"> nanoMB: Oxford Nanopore 16S Pipeline (Container-First)</h1>
<p>This workflow processes full-length Oxford Nanopore 16S amplicon reads from raw POD5 files through demultiplexing, consensus polishing, and taxonomic assignment. It is optimized for the Swan HPC environment and runs entirely from Apptainer containers.</p>
<hr>
<section id="overview" class="level2">
<h2 class="anchored" data-anchor-id="overview">üß≠ Overview</h2>
<section id="workflow-summary" class="level3">
<h3 class="anchored" data-anchor-id="workflow-summary">Workflow summary</h3>
<table class="caption-top table">
<colgroup>
<col style="width: 21%">
<col style="width: 18%">
<col style="width: 37%">
<col style="width: 21%">
</colgroup>
<thead>
<tr class="header">
<th>Stage</th>
<th>Tool</th>
<th>Container</th>
<th>Notes</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Basecalling</td>
<td><strong>Dorado</strong></td>
<td><code>docker:nanoporetech/dorado:latest</code></td>
<td>GPU-based; uses ONT R10.4.1 SUP model</td>
</tr>
<tr class="even">
<td>Demultiplexing</td>
<td><strong>Dorado demux</strong></td>
<td><code>docker:nanoporetech/dorado:latest</code></td>
<td>Uses sample sheets under <code>samples/</code></td>
</tr>
<tr class="odd">
<td>Trimming</td>
<td><strong>Dorado trim</strong></td>
<td><code>docker:nanoporetech/dorado:latest</code></td>
<td>Generates per-sample FASTQs</td>
</tr>
<tr class="even">
<td>QC / Filtering</td>
<td><strong>Fastcat</strong>, <strong>NanoPlot</strong></td>
<td><code>docker:aliciamrich/nanomb</code></td>
<td>Length &amp; Q-score filters, QC plots</td>
</tr>
<tr class="odd">
<td>Clustering</td>
<td><strong>isONclust3</strong></td>
<td><code>docker:aliciamrich/nanomb</code></td>
<td>ONT-specific clustering of reads</td>
</tr>
<tr class="even">
<td>Consensus</td>
<td><strong>SPOA</strong></td>
<td><code>docker:aliciamrich/nanomb</code></td>
<td>Builds per-cluster consensus sequences</td>
</tr>
<tr class="odd">
<td>Pooling / OTUs</td>
<td><strong>VSEARCH</strong></td>
<td><code>docker:aliciamrich/nanomb</code></td>
<td>Dereplication &amp; clustering at 99 %/97 %</td>
</tr>
<tr class="even">
<td>Polishing</td>
<td><strong>Racon √ó2</strong>, <strong>Medaka</strong></td>
<td><code>docker:aliciamrich/nanombgpu</code></td>
<td>GPU polish to high-accuracy OTUs</td>
</tr>
<tr class="odd">
<td>Taxonomy / Tree</td>
<td><strong>VSEARCH SINTAX</strong>, <strong>MAFFT</strong>, <strong>IQ-TREE 2</strong></td>
<td><code>docker:aliciamrich/nanomb</code></td>
<td>Classify OTUs and infer phylogeny</td>
</tr>
<tr class="even">
<td>ASV option</td>
<td><strong>nanoASV</strong></td>
<td><code>docker:aliciamrich/nanoasv</code></td>
<td><em>Third-party workflow disabled by default for ONT kits</em></td>
</tr>
</tbody>
</table>
<hr>
</section>
</section>
<section id="directory-layout" class="level2">
<h2 class="anchored" data-anchor-id="directory-layout">‚öôÔ∏è Directory Layout</h2>
<p>Data and outputs are resolved from environment variables defined by<br>
<a href="profiles/hcc/env_setup.sh"><code>profiles/hcc/env_setup.sh</code></a>.</p>
<table class="caption-top table">
<colgroup>
<col style="width: 30%">
<col style="width: 41%">
<col style="width: 27%">
</colgroup>
<thead>
<tr class="header">
<th>Variable</th>
<th>Typical path</th>
<th>Purpose</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>PROJ_ROOT</code></td>
<td><code>/mnt/nrdstor/richlab/shared/nanomb</code></td>
<td>Shared repo with Snakefile</td>
</tr>
<tr class="even">
<td><code>WORK</code></td>
<td><code>/work/richlab/$USER</code></td>
<td>Your Swan work directory</td>
</tr>
<tr class="odd">
<td><code>NRDSTOR</code></td>
<td><code>/mnt/nrdstor/richlab/$USER</code></td>
<td>Long-term storage</td>
</tr>
<tr class="even">
<td><code>OUT_ROOT</code></td>
<td><code>$WORK/datasets/16s</code></td>
<td>Root for all datasets</td>
</tr>
<tr class="odd">
<td><code>sampleset</code></td>
<td>e.g.&nbsp;<code>loris</code></td>
<td>Sample collection name</td>
</tr>
<tr class="even">
<td><code>dataset</code></td>
<td>e.g.&nbsp;<code>culi</code></td>
<td>Logical dataset under a sampleset</td>
</tr>
</tbody>
</table>
<p>Pipeline outputs land under:</p>
<pre><code>$WORK/datasets/16s///
‚îÇ
‚îú‚îÄ‚îÄ raw//                # demuxed + trimmed per-run reads
‚îú‚îÄ‚îÄ tmp/                 # intermediate results
‚îú‚îÄ‚îÄ qc/                  # QC reports
‚îú‚îÄ‚îÄ otu/                 # clustered OTUs, taxonomy, tree
‚îî‚îÄ‚îÄ asv/                 # ASV branch (placeholder if disabled)</code></pre>
<hr>
</section>
<section id="running-on-swan" class="level2">
<h2 class="anchored" data-anchor-id="running-on-swan">üöÄ Running on Swan</h2>
<section id="log-into-your-swan-account" class="level3">
<h3 class="anchored" data-anchor-id="log-into-your-swan-account">0. Log into your swan account</h3>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ssh</span> username@swan.unl.edu</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><code>enter your password and duo authentication</code></p>
</section>
<section id="create-snakemake-conda-environment" class="level3">
<h3 class="anchored" data-anchor-id="create-snakemake-conda-environment">0.5. Create Snakemake conda environment</h3>
<p>You need to access the snakemake module through a <a href="https://hcc.unl.edu/docs/applications/user_software/using_anaconda_package_manager/">virtual conda environment</a>, but only on the first use. After that, you can load the same environment each time (see next).</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load anaconda</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> create <span class="at">-c</span> conda-forge <span class="at">-c</span> bioconda <span class="at">-c</span> nodefaults <span class="at">-p</span> <span class="va">$NRDSTOR</span>/snakemake snakemake</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate <span class="va">$NRDSTOR</span>/snakemake</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="load-your-environment" class="level3">
<h3 class="anchored" data-anchor-id="load-your-environment">1. Load your environment</h3>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> /mnt/nrdstor/richlab/shared/nanomb</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> profiles/hcc/env_setup.sh</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="dry-run-the-workflow" class="level3">
<h3 class="anchored" data-anchor-id="dry-run-the-workflow">2. Dry-run the workflow</h3>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load anaconda</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate <span class="va">$NRDSTOR</span>/snakemake</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="ex">snakemake</span> <span class="at">-np</span> <span class="at">--profile</span> profiles/hcc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="submit-to-slurm" class="level3">
<h3 class="anchored" data-anchor-id="submit-to-slurm">3. Submit to SLURM</h3>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">sbatch</span> scripts/run_nanomb.sh</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The orchestrator job stays light; each rule submits its own SLURM job via the profile.</p>
<p>Logs are written to <code>/mnt/nrdstor/richlab/shared/nanomb/.snakemake/slurm_logs/.</code></p>
<p>‚∏ª</p>
</section>
</section>
<section id="configuration" class="level2">
<h2 class="anchored" data-anchor-id="configuration">üß© Configuration</h2>
<p>All parameters live in <a href="config/config.yaml">config/config.yaml</a>.</p>
<section id="key-sections" class="level3">
<h3 class="anchored" data-anchor-id="key-sections">Key sections:</h3>
<div class="sourceCode" id="cb7"><pre class="sourceCode yaml code-with-copy"><code class="sourceCode yaml"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Identify this dataset</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sampleset</span><span class="kw">:</span><span class="at"> loris</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dataset</span><span class="kw">:</span><span class="at">   culi</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="fu">runs</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="at">hdz1</span><span class="kw">,</span><span class="at"> hdz2</span><span class="kw">,</span><span class="at"> ...</span><span class="kw">,</span><span class="at"> hdz19</span><span class="kw">]</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Data root</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="fu">out_root</span><span class="kw">:</span><span class="at"> </span><span class="st">"$WORK/datasets/16s"</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a><span class="co"># References</span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a><span class="fu">itgdb_udb</span><span class="kw">:</span><span class="at">   </span><span class="st">"/mnt/nrdstor/richlab/shared/databases/itgdb_16s/taxa_itgdb.udb"</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a><span class="fu">silva_fasta</span><span class="kw">:</span><span class="at"> </span><span class="st">"/mnt/nrdstor/richlab/shared/databases/silva138/SILVA_138.2_SSURef_NR99_tax_silva.fasta"</span></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Disable ASV branch (recommended for ONT)</span></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a><span class="fu">asv_method</span><span class="kw">:</span><span class="at"> none</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>‚∏ª</p>
</section>
</section>
<section id="notes-known-issues" class="level2">
<h2 class="anchored" data-anchor-id="notes-known-issues">üß† Notes &amp; Known Issues</h2>
<ul>
<li><code>spoa_consensus</code> requires temporary files with .fastq extension (handled automatically in the current Snakefile).</li>
<li><code>nanoASV</code> often fails with ONT kits that already include Dorado trimming. The branch is disabled by default via asv_method: none.</li>
<li>The ILP solver (CBC) messages during job scheduling are normal; Snakemake uses them to decide job ordering.</li>
<li>If you add new containers, update paths in config.yaml under container_*.</li>
<li>Re-run a failed rule safely with:</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="ex">snakemake</span> <span class="op">&lt;</span>rule<span class="op">&gt;</span> --profile profiles/hcc <span class="at">--rerun-incomplete</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>‚∏ª</p>
</section>
<section id="typical-output-artifacts" class="level2">
<h2 class="anchored" data-anchor-id="typical-output-artifacts">üß™ Typical Output Artifacts</h2>
<table class="caption-top table">
<colgroup>
<col style="width: 25%">
<col style="width: 74%">
</colgroup>
<thead>
<tr class="header">
<th>File</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>otu/otus_centroids_99.fasta</td>
<td>Primary OTU centroids (99 % identity)</td>
</tr>
<tr class="even">
<td>otu/otus_centroids_97.fasta</td>
<td>Secondary OTU centroids (97 % identity)</td>
</tr>
<tr class="odd">
<td>otu/otus_taxonomy.sintax</td>
<td>Taxonomic annotations from VSEARCH SINTAX</td>
</tr>
<tr class="even">
<td>otu/otu_references_aligned.fasta</td>
<td>Multiple-sequence alignment (MAFFT)</td>
</tr>
<tr class="odd">
<td>otu/otu_tree.treefile</td>
<td>Maximum-likelihood phylogeny (IQ-TREE 2)</td>
</tr>
<tr class="even">
<td>otu/otu_table_merged.tsv</td>
<td>Merged per-sample OTU abundance table</td>
</tr>
<tr class="odd">
<td>qc/fastcat_file_summary.tsv</td>
<td>Per-file QC summary from Fastcat</td>
</tr>
<tr class="even">
<td>qc/fastcat_read_summary.tsv</td>
<td>Per-read QC summary from Fastcat</td>
</tr>
<tr class="odd">
<td>qc/nanoplot/</td>
<td>NanoPlot read-length and quality plots</td>
</tr>
<tr class="even">
<td>tmp/polished/polished_otus.fasta</td>
<td>Final Medaka-polished OTUs</td>
</tr>
<tr class="odd">
<td>asv/nanoasv/phyloseq.RData</td>
<td>Placeholder (empty if asv_method: none)</td>
</tr>
</tbody>
</table>
<p>‚∏ª</p>
</section>
<section id="containers" class="level2">
<h2 class="anchored" data-anchor-id="containers">üß∞ Containers</h2>
<p>All stages run inside Apptainer containers (bind mounts set in the profile):</p>
<table class="caption-top table">
<colgroup>
<col style="width: 30%">
<col style="width: 15%">
<col style="width: 53%">
</colgroup>
<thead>
<tr class="header">
<th>Container</th>
<th>Path</th>
<th>Purpose / Key Tools</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>nanomb.sif</td>
<td>/mnt/nrdstor/richlab/shared/containers/nanomb.sif</td>
<td>CPU environment (Fastcat, VSEARCH, isONclust3, SPOA, MAFFT, IQ-TREE 2, etc.)</td>
</tr>
<tr class="even">
<td>nanombgpu.sif</td>
<td>/mnt/nrdstor/richlab/shared/containers/nanombgpu.sif</td>
<td>GPU tools (Racon √ó 2, Medaka polish)</td>
</tr>
<tr class="odd">
<td>dorado.sif</td>
<td>/mnt/nrdstor/richlab/shared/containers/dorado.sif</td>
<td>GPU basecalling, demultiplexing, and trimming with ONT Dorado</td>
</tr>
<tr class="even">
<td>nanoasv.sif</td>
<td>/mnt/nrdstor/richlab/shared/containers/nanoasv.sif</td>
<td>Optional ASV generation (disabled by default)</td>
</tr>
</tbody>
</table>
<p>‚∏ª</p>
</section>
<section id="citation" class="level2">
<h2 class="anchored" data-anchor-id="citation">üßæ Citation</h2>
<p>If you use this workflow in analyses or teaching:</p>
<p>Rich Lab (2025). nanoMB: container-based ONT 16S processing pipeline.</p>
<p>‚∏ª</p>
<dl>
<dt>Author</dt>
<dd>
Alicia M. Rich, PhD
</dd>
<dt>License</dt>
<dd>
MIT (open for teaching and research use)
</dd>
</dl>
<p>‚∏ª</p>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "Óßã";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>